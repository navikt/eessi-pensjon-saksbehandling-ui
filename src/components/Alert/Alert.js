import React from 'react'
import PT from 'prop-types'
import classNames from 'classnames'

import { connect, bindActionCreators } from 'store'
import Icons from 'components/Icons'
import { AlertStripe } from 'components/Nav'
import * as alertActions from 'actions/alert'
import { getDisplayName } from 'utils/displayName'

import './Alert.css'

export const mapStateToProps = (state) => {
  return {
    clientErrorStatus: state.alert.clientErrorStatus,
    clientErrorMessage: state.alert.clientErrorMessage,
    serverErrorMessage: state.alert.serverErrorMessage,
    error: state.alert.error
  }
}

export const mapDispatchToProps = (dispatch) => {
  return { actions: bindActionCreators(alertActions, dispatch) }
}

export const errorTypes = {
  OK: 'suksess',
  ERROR: 'feil',
  WARNING: 'advarsel'
}

export const Alert = (props) => {
  const { actions, className, clientErrorStatus, clientErrorMessage, error, fixed, t, type, serverErrorMessage } = props

  const onClientClear = () => {
    actions.clientClear()
  }

  const printError = (error) => {
    let errorMessage = []
    if (error.status) {
      errorMessage.push(error.status)
    }
    if (error.message) {
      errorMessage.push(error.message)
    }
    if (error.serverMessage) {
      errorMessage.push(error.serverMessage)
    }
    if (error.uuid) {
      errorMessage.push(error.uuid)
    }
    return errorMessage.join(' ')
  }

  if (type === 'server') {
    if (!serverErrorMessage) {
      return null
    }

    let message = t(serverErrorMessage)
    if (error) {
      message += ': ' + printError(error)
    }
    return <AlertStripe
      className={classNames('c-ui-alert', 'server', className)} type={errorTypes.ERROR}>
      {message}
      <Icons className='closeIcon' size='1x' kind='solidclose' onClick={onClientClear} />
    </AlertStripe>
  }

  if (!clientErrorMessage) {
    return null
  }

  const separatorIndex = clientErrorMessage.lastIndexOf('|')
  let message

  if (separatorIndex >= 0) {
    message = t(clientErrorMessage.substring(0, separatorIndex)) + ': ' + clientErrorMessage.substring(separatorIndex + 1)
  } else {
    message = t(clientErrorMessage)
  }

  if (error) {
    message += ': ' + printError(error)
  }

  return <AlertStripe
    className={classNames(className, 'c-ui-alert', { fixed: fixed || true })}
    type={errorTypes[clientErrorStatus]}>
    {message}
    <Icons className='closeIcon' size='1x' kind='solidclose' onClick={onClientClear} />
  </AlertStripe>
}

Alert.propTypes = {
  actions: PT.object.isRequired,
  className: PT.string,
  clientErrorStatus: PT.string,
  clientErrorMessage: PT.string,
  error: PT.object,
  fixed: PT.bool,
  t: PT.func.isRequired,
  type: PT.string.isRequired,
  serverErrorMessage: PT.string
}

const ConnectedAlert = connect(mapStateToProps, mapDispatchToProps)(Alert)
ConnectedAlert.displayName = `Connect(${getDisplayName((Alert))})`
export default ConnectedAlert
