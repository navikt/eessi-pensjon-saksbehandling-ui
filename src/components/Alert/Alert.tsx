import FilledRemoveCircle from 'assets/icons/filled-version-remove-circle'
import classNames from 'classnames'
import { fadeIn } from 'components/keyframes'
import { AlertError, AlertStatus } from 'declarations/components'
import { AlertErrorPropType } from 'declarations/components.pt'
import _ from 'lodash'
import AlertStripe, { AlertStripeType } from 'nav-frontend-alertstriper'
import PT from 'prop-types'
import React from 'react'
import styled from 'styled-components'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export type AlertStatusClasses = {[status in AlertStatus]: AlertStripeType}

export type AlertType = 'client' | 'server'

export const AlertDiv = styled(AlertStripe)`
  opacity: 0;
  animation: ${fadeIn} 1s forwards;
  position: relative;

  .alertstripe .alertstripe__tekst {
     flex: auto !important; /* because IE11 */
  }
  &.fixed {
    position: fixed;
    top: 0.25rem;
    left: 50%;
    transform: translateX(-50%);
    z-index: 10;
    min-width: 400px;
  }
  &.type-server {
    border-radius: 0px !important;
    border: 0px !important;
    .alertstripe__tekst {
      display: flex !important;
      justify-content: space-between;
      max-width: none !important;
    }
  }
`
export const CloseIcon = styled(FilledRemoveCircle)`
  position: absolute;
  top: 0.25rem;
  right: 0.25rem;
  cursor: pointer;
`

export interface AlertProps {
  className ?: string
  error?: AlertError | string
  fixed?: boolean
  message?: JSX.Element | string
  onClose?: () => void
  status?: AlertStatus
  type?: AlertType
}

export const errorTypes: AlertStatusClasses = {
  OK: 'suksess',
  ERROR: 'feil',
  WARNING: 'advarsel'
}

export const Alert: React.FC<AlertProps> = ({
  className, error, fixed, message, onClose, status = 'ERROR', type
}: AlertProps): JSX.Element => {
  let _message: JSX.Element | string | undefined = message

  const onCloseIconClicked = (): void => {
    if (_.isFunction(onClose)) {
      onClose()
    }
  }

  const printError = (error: AlertError | string): string => {
    const errorMessage: Array<JSX.Element | string> = []
    if (_.isString(error)) {
      return error
    }
    if (error.message) {
      errorMessage.push(error.message)
    }
    if (error.error) {
      errorMessage.push(error.error)
    }
    if (error.uuid) {
      errorMessage.push(error.uuid)
    }
    return errorMessage.join(' - ')
  }

  if (!_message) {
    return <div />
  }

  if (!_.includes(['client', 'server'], type)) {
    console.error('Invalid alert type: ' + type)
    return <div />
  }

  if (!_.includes(Object.keys(errorTypes), status)) {
    console.error('Invalid alert status: ' + status)
    return <div />
  }

  if (!_.isEmpty(error)) {
    _message += ': ' + printError(error!)
  }

  const _fixed: boolean = _.isNil(fixed) ? type === 'client' : fixed
  return (
    <AlertDiv
      className={classNames(
        'type-' + type,
        'status-' + status,
        className,
        { fixed: _fixed }
      )}
      type={errorTypes[status]}
    >
      {_message}
      {onClose && (
        <CloseIcon
          data-test-id='c-alert__close-icon'
          onClick={onCloseIconClicked}
        />
      )}
    </AlertDiv>
  )
}

Alert.propTypes = {
  className: PT.string,
  error: PT.oneOfType([AlertErrorPropType, PT.string]),
  fixed: PT.bool,
  message: PT.oneOfType([PT.string, PT.element]),
  onClose: PT.func,
  status: PT.oneOf(['OK', 'ERROR', 'WARNING']),
  type: PT.oneOf(['client', 'server'])
}

Alert.displayName = 'Alert'
export default Alert
