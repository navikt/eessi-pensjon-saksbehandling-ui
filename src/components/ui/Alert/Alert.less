.c-ui-alert {

    height: 56px;
    min-height: 56px;
    min-width: 300px;
    opacity: 0;
    animation: c-ui-alert-fadeIn 0.5s forwards;

    &.fixed {
        position: fixed;
        top: 4px;
        left: 50%;
        transform: translateX(-50%);
        z-index: 10;
    }

    .closeIcon {
        position: absolute;
        top: 3px;
        right: 3px;
        cursor: pointer;
    }

    &.server {

      .alertstripe__tekst {
        display: flex !important;
        justify-content: space-between;
        max-width: none !important;
      }
    }
}

@keyframes c-ui-alert-fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
