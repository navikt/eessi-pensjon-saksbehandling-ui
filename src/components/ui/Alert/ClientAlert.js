import React, { Component } from 'react';
import PT from 'prop-types';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';
import classNames from 'classnames';
import { bindActionCreators }  from 'redux';

import * as alertActions from '../../../actions/alert';

import * as Nav from '../Nav';
import './ClientAlert.css'

const mapStateToProps = (state) => {
    return {
        clientErrorStatus  : state.alert.clientErrorStatus,
        clientErrorMessage : state.alert.clientErrorMessage
    }
};

const mapDispatchToProps = (dispatch) => {
    return {actions: bindActionCreators(Object.assign({}, alertActions), dispatch)};
};

class ClientAlert extends Component {

    state = {
        status  : undefined,
        message : undefined,
        timeout : undefined
    }

    clear () {
        const { clientErrorStatus, actions } = this.props;

        // after 5 seconds, if the client alert is still not an error, then clear it
        if (clientErrorStatus !== 'ERROR') {
            actions.clientClear();
        }
    }

    handleTimeouts() {

        const { clientErrorStatus, clientErrorMessage } = this.props;

        if (clientErrorStatus !== this.state.status || clientErrorMessage !== this.state.message) {
            if (this.state.timeout) {
                clearTimeout(this.state.timeout);
            }

            let timeout = undefined;
            let self = this;

            if (clientErrorStatus === 'OK') {
                timeout = setTimeout(() => {
                    self.clear();
                }, 5000);
            }

            this.setState({
                timeout: timeout,
                status: clientErrorStatus,
                message: clientErrorMessage
            });
        }
    }

    render () {

        let { t, clientErrorStatus, clientErrorMessage, permanent, className } = this.props;

        if (!clientErrorMessage) {
                return permanent ? <div style={{height: '75px'}}
                className={classNames(className, 'p-4', 'm-4')}>&nbsp;</div> : null
        }

        this.handleTimeouts();

        let message = undefined;
        let separatorIndex = clientErrorMessage.lastIndexOf('|');

        if (separatorIndex >= 0) {
            message = t(clientErrorMessage.substring(0, separatorIndex)) + ': ' + clientErrorMessage.substring(separatorIndex + 1);
        } else {
            message = t(clientErrorMessage);
        }
        return <Nav.AlertStripe className={classNames(className, 'clientAlert', 'm-4', {'toFade' : clientErrorStatus === 'OK'})}
            type={clientErrorStatus === 'OK' ? 'suksess' : 'advarsel'}>
            {message}
        </Nav.AlertStripe>;
    }
}

ClientAlert.propTypes = {
    t                  : PT.func.isRequired,
    permanent          : PT.bool,
    clientErrorStatus  : PT.string,
    clientErrorMessage : PT.string,
    actions            : PT.object.isRequired,
    className          : PT.string
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(
    translate()(ClientAlert)
);
