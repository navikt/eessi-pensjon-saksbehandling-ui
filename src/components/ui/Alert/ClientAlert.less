.c-ui-clientAlert {

    height: 60px;
    min-height: 60px;
    min-width: 300px;
    opacity: 0;
    animation: c-ui-clientAlert-fadeIn 0.5s forwards;

    &.fixed {
        position: fixed;
        top: 3px;
        left: 50%;
        transform: translateX(-50%);
        z-index: 10;
    }

    .closeIcon {
        position: absolute;
        top: 3px;
        right: 3px;
        cursor: pointer;
    }
}

@keyframes c-ui-clientAlert-fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
