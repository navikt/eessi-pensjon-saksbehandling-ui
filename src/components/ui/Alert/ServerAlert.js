import React, { Component } from 'react'
import PT from 'prop-types'
import { connect } from 'react-redux'
import { withNamespaces } from 'react-i18next'

import * as Nav from '../Nav'

const mapStateToProps = (state) => {
  return {
    serverErrorMessage: state.alert.serverErrorMessage
  }
}

class ServerAlert extends Component {
  render () {
    let { t, serverErrorMessage } = this.props

    return serverErrorMessage ? <Nav.AlertStripe
      className='c-ui-serverAlert' type='advarsel' solid>
      {t(serverErrorMessage)}
    </Nav.AlertStripe> : null
  }
}

ServerAlert.propTypes = {
  t: PT.func.isRequired,
  serverErrorMessage: PT.string
}

export default connect(
  mapStateToProps
)(
  withNamespaces()(ServerAlert)
)
