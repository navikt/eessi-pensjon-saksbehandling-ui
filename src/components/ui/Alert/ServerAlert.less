.c-ui-serverAlert {
  .alertstripe__tekst {
    margin-top: -0.75rem;
  }
}
