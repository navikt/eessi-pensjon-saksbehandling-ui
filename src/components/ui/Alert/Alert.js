import React, { Component } from 'react'
import PT from 'prop-types'
import { withTranslation } from 'react-i18next'
import classNames from 'classnames'
import { connect, bindActionCreators } from 'store'

import Icons from '../Icons'
import * as Nav from '../Nav'

import * as alertActions from '../../../actions/alert'

import './Alert.css'
import { getDisplayName } from '../../../utils/displayName'

export const mapStateToProps = (state) => {
  return {
    clientErrorStatus: state.alert.clientErrorStatus,
    clientErrorMessage: state.alert.clientErrorMessage,
    serverErrorMessage: state.alert.serverErrorMessage,
    uuid: state.alert.uuid
  }
}

export const mapDispatchToProps = (dispatch) => {
  return { actions: bindActionCreators(Object.assign({}, alertActions), dispatch) }
}

export const errorTypes = {
  OK: 'suksess',
  ERROR: 'feil',
  WARNING: 'advarsel'
}

export class Alert extends Component {
  clientClear () {
    const { actions } = this.props
    actions.clientClear()
  }

  render () {
    let { t, className, type, clientErrorStatus, clientErrorMessage, serverErrorMessage, uuid, fixed } = this.props

    if (type === 'server') {
      return serverErrorMessage ? <Nav.AlertStripe
        className={classNames('c-ui-alert', 'server', className)} type='feil'>
        {t(serverErrorMessage)}
        {uuid}
        <Icons className='closeIcon' size='1x' kind='solidclose' onClick={this.clientClear.bind(this)} />
      </Nav.AlertStripe> : null
    }

    if (!clientErrorMessage) {
      return null
    }

    let separatorIndex = clientErrorMessage.lastIndexOf('|')
    let _message
    let _fixed = fixed || true

    if (separatorIndex >= 0) {
      _message = t(clientErrorMessage.substring(0, separatorIndex)) + ': ' + clientErrorMessage.substring(separatorIndex + 1)
    } else {
      _message = t(clientErrorMessage)
    }
    return <Nav.AlertStripe
      className={classNames(className, 'c-ui-alert', { fixed: _fixed })}
      type={errorTypes[clientErrorStatus]}>
      {_message}
      {uuid}
      <Icons className='closeIcon' size='1x' kind='solidclose' onClick={this.clientClear.bind(this)} />
    </Nav.AlertStripe>
  }
}

Alert.propTypes = {
  t: PT.func.isRequired,
  actions: PT.object.isRequired,
  className: PT.string,
  type: PT.string.isRequired,
  fixed: PT.bool,
  clientErrorStatus: PT.string,
  clientErrorMessage: PT.string,
  serverErrorMessage: PT.string,
  uuid: PT.string
}

const ConnectedAlert = connect(
  mapStateToProps,
  mapDispatchToProps
)(
  withTranslation()(Alert)
)

ConnectedAlert.displayName = `Connect(${getDisplayName((
  withTranslation()(Alert)
))})`

export default ConnectedAlert
