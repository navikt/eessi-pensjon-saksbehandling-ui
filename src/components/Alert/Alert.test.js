import React from 'react'
import { connect, bindActionCreators } from 'store'

import { Alert, mapStateToProps } from './Alert'
import * as reducers from '../../../reducers'
import * as alertActions from '../../../actions/alert'

const t = jest.fn((translationString) => { return translationString })

const reducer = combineReducers({
  ...reducers
})

const mapDispatchToProps = (dispatch) => {
  return { actions: bindActionCreators(Object.assign({}, alertActions), dispatch) }
}

describe('components/Alert/Alert', () => {
  let store, wrapper, ConnectedAlert
  const initialAlertState = {
    alert: {
      clientErrorMessage: 'mockClientErrorMessage',
      clientErrorStatus: 'mockClientErrorStatus',
      serverErrorMessage: 'mockServerErrorMessage',
      uuid: 'MockUuid'
    }
  }

  beforeEach(() => {
    store = createStore(reducer, initialAlertState)
    ConnectedAlert = connect(mapStateToProps, mapDispatchToProps)(Alert)
  })

  it('Alert renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Alert renders as server', () => {
    wrapper = shallow(<ConnectedAlert t={t} store={store} type='server' />).dive()
    expect(wrapper.render().text()).toEqual('mockServerErrorMessageMockUuid')
  })

  it('Alert renders as client', () => {
    wrapper = shallow(<ConnectedAlert t={t} store={store} type='client' />).dive()
    expect(wrapper.render().text()).toEqual('mockClientErrorMessageMockUuid')
  })

  it('Alert close button works', async (done) => {
    wrapper = shallow(<ConnectedAlert t={t} store={store} type='client' />).dive()
    expect(wrapper.render().text()).toEqual('mockClientErrorMessageMockUuid')
    wrapper.instance().clientClear()
    await new Promise(resolve => {
      setTimeout(() => {
        expect(wrapper.render().text()).toEqual('')
        done()
      }, 500)
    })
  })
})
