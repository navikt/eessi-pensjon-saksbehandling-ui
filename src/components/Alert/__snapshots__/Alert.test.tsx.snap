// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Alert/Alert Renders 1`] = `
<Alert
  error={Object {}}
  message="mockErrorMessage"
  onClose={[MockFunction]}
  status="OK"
  type="server"
>
  <AlertStripe
    className="c-alert c-alert__type-server c-alert__status-OK"
    form="default"
    size="1.5em"
    type="suksess"
  >
    <div
      className="alertstripe c-alert c-alert__type-server c-alert__status-OK alertstripe--suksess"
    >
      <span
        className="alertstripe__ikon"
      >
        <span
          className="sr-only"
        >
          suksess
        </span>
        <Icon
          kind="ok-sirkel-fyll"
          size="1.5em"
        >
          <svg
            focusable="false"
            height="1.5em"
            kind="ok-sirkel-fyll"
            viewBox="0 0 24 24"
            width="1.5em"
          >
            <g
              fill="none"
              fillRule="nonzero"
            >
              <path
                d="M12 0C5.383 0 0 5.384 0 12s5.383 12 12 12c6.616 0 12-5.384 12-12S18.616 0 12 0z"
                fill="#1C6937"
              />
              <path
                d="M9.64 14.441l6.46-5.839a.997.997 0 0 1 1.376.044.923.923 0 0 1-.046 1.334l-7.15 6.464a.993.993 0 0 1-.662.252.992.992 0 0 1-.69-.276l-2.382-2.308a.923.923 0 0 1 0-1.334.997.997 0 0 1 1.377 0l1.717 1.663z"
                fill="#FFF"
              />
            </g>
          </svg>
        </Icon>
      </span>
      <Normaltekst
        className="alertstripe__tekst"
        tag="div"
      >
        <TypografiBase
          className="alertstripe__tekst"
          tag="div"
          type="normaltekst"
        >
          <div
            className="typo-normal alertstripe__tekst"
          >
            mockErrorMessage
            <Icons
              className="closeIcon"
              kind="solidclose"
              onClick={[Function]}
            >
              <FontAwesomeIcon
                border={false}
                className="closeIcon"
                fixedWidth={false}
                flip={null}
                icon={
                  Object {
                    "icon": Array [
                      512,
                      512,
                      Array [],
                      "f057",
                      "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z",
                    ],
                    "iconName": "times-circle",
                    "prefix": "fas",
                  }
                }
                inverse={false}
                kind="solidclose"
                listItem={false}
                mask={null}
                onClick={[Function]}
                pull={null}
                pulse={false}
                rotation={null}
                size="1x"
                spin={false}
                swapOpacity={false}
                symbol={false}
                title=""
                transform={null}
              >
                <svg
                  aria-hidden="true"
                  className="svg-inline--fa fa-times-circle fa-w-16 fa-1x closeIcon"
                  data-icon="times-circle"
                  data-prefix="fas"
                  focusable="false"
                  kind="solidclose"
                  onClick={[Function]}
                  role="img"
                  style={Object {}}
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z"
                    fill="currentColor"
                    style={Object {}}
                  />
                </svg>
              </FontAwesomeIcon>
            </Icons>
          </div>
        </TypografiBase>
      </Normaltekst>
    </div>
  </AlertStripe>
</Alert>
`;
